### PR TITLE
Pool (#15879)

### PR DESCRIPTION
**Pull request by mahmodKhaled**
- created pool function in TensorFlow frontend at nn.py file
- implemented the pool function in ivy functional in layers.pi file with a docstring

### CODE REVIEW

**Comment by ZiadAmerr** [comment]
Hi @mahmodKhaled, thanks a lot for your contribution. Feel free to request a review once you are done with the PR. Also, make sure that the linting tests are passing locally. You can do so by executing `flake8 foo.py` for each file you have modified and fix all the errors related to your modifications. If the errors are due to other files or parts that you haven't modified, feel free to ignore them. :)

**Comment by mahmodKhaled** [comment]
hi @ZiadAmerr, I have passed 62 test cases, but failed in 5, please I want a code review here, to know what are the next steps. Thank you in advance.

**Comment by ZiadAmerr** [comment]
Amazing work! Can we implement the tests needed for testing our implementation in their designated files? :)

**Comment by mahmodKhaled** [comment]
Hi @ZiadAmerr I managed to add a test case in the file test_nn.py, and passed more test checks, there a two left, I need to know where is the problem to debug the code, and if there is extra code I need to add in some other file, or if my implementation is not correct. I'm now 65 tests passed and 2 failed. Thank you again.

**Comment by ZiadAmerr** [comment]
Are you sure that the tests are passing locally? There seems to be an error from your testing function in this specific line:
`(input_dtype, x, ksize, strides, padding), data_format = x_k_s_p_df`

You can check out the error from [here](https://github.com/unifyai/ivy/actions/runs/5088514841/jobs/9145013532#step:3:1)

**Review by ZiadAmerr — CHANGES_REQUESTED** [review]
Hey @mahmodKhaled, amazing work! Just a couple changes and we can proceed with the merge :)

**Comment by mahmodKhaled** [comment]
Hi @ZiadAmerr, I have made major changes here, i will list them now:
1. move the implementation of pool function to the right designated file as you told to me, and hence edited the implementation of the function. now the function is located inside experimental folder in layers.py file.
2. I have edited the line in the test case function that makes the error, make many changes to it, and now everything works fine
3. corrected the parameters of the pool function, because they were wrong.
4. Finally, I have passed all test checks except 1, that test the function, but it give me this error and I don't understand why, because i revised no. of parameters passed and I have made sure that they are correct no extra values passed, but I still have the following error
```
 TypeError: pool() takes 3 positional arguments but 4 positional arguments (and 5 keyword-only arguments) were given
 ```
 I would like to have some help here, and referencing that you told me to test the function locally, I can't test the code in my Pycharm, it gives more a lot of import error, so I didn't make to test them locally, so I depend a lot on the output of Github Action. Thanks in advance.

**Comment by mahmodKhaled** [comment]
Hi @ZiadAmerr, I have solved the above error, and now a new 2 errors occur one not related to my function, and another is related to this error ``` RecursionError: maximum recursion depth exceeded in comparison ```, please help me figure out what is the problem is, or does the problem is not from my code ?. Thanks in advance.

**Comment by ZiadAmerr** [comment]
closes #15879

**Review by ZiadAmerr — APPROVED** [review]
Amazing work, @mahmodKhaled! Will do a couple reruns on the tests and we are ready to merge :)

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py [hunk 1]
```diff
@@ -1490,3 +1490,66 @@ def test_tensorflow_avg_pool1d(
...
+        pooling_type,
+        data_format,
+        dims + 2,
+    )
+
+
+# pool
+@handle_frontend_test(
+    fn_tree="tensorflow.nn.pool",
+    x_k_s_p_df=_pool_args(),
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py:1520]: Can you test whether this works with the helper `get_arrays_for_pooling(...)` as used previously in [here](https://github.com/unifyai/ivy/pull/15884/files#diff-259a7991f1af0d34dea0190b1b676d36560034f3d2075c8337f6d6c7c08225e2R1468-R1470)

Comment (mahmodKhaled) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py:1520]: Hi, @ZiadAmerr, it seems like this function ```get_arrays_for_pooling(...)``` doesn't exist in my current version of the repo.

Comment (ZiadAmerr) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py:1520]: Sorry, it is `helpers.arrays_for_pooling(...)`.

You can find it in the link of my previous comment

**Comment by mahmodKhaled** [comment]
Hi @ZiadAmerr, I have applied the changes you requested above, and my pool function successfully throws the right exception if there is no pooling type given, I assumed the following:
1- if we used the helper testing function as you mentioned above, so this function doesn't return any parameter that specifies the pooling type, so I put it to ```UNKNOWN``` value. and my function throws the right exception as in this case the user didn't provide any pooling type.

further improvements: I can hardcode the ```AVG``` or ```MAX``` pooling type value, to test if it is works fine, if you want this i can try

I'm waiting for your code review, and to hear for any wanted changes to apply.

File: ivy/functional/frontends/tensorflow/nn.py [hunk 2]
```diff
@@ -477,3 +477,44 @@ def avg_pool3d(input, ksize, strides, padding, data_format="NDHWC", name=None):
...
+
+@to_ivy_arrays_and_back
+def pool(
+    input,
+    window_shape,
+    pooling_type,
+    strides=None,
+    padding="VALID",
+    data_format=None,
+    dilations=None,
```
Comment (ZiadAmerr) [inline ivy/functional/frontends/tensorflow/nn.py:490]: This parameter hasn't been tested.

File: ivy/functional/frontends/tensorflow/nn.py [hunk 3]
```diff
@@ -477,3 +477,44 @@ def avg_pool3d(input, ksize, strides, padding, data_format="NDHWC", name=None):
...
     return ivy.avg_pool1d(input, ksize, strides, padding, data_format=data_format)
+
+
+@to_ivy_arrays_and_back
+def pool(
+    input,
+    window_shape,
+    pooling_type,
+    strides=None,
+    padding="VALID",
```
Comment (ZiadAmerr) [inline ivy/functional/frontends/tensorflow/nn.py:488]: This parameter hasn't been tested. You can use `st.one_of(...)`.

**Review by ZiadAmerr — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/tensorflow/experimental/layers.py [hunk 4]
```diff
@@ -444,7 +444,33 @@ def avg_pool3d(
     return res
 
 
-@with_supported_dtypes({"2.12.0 and below": ("float32", "float64")}, backend_version)
+@with_unsupported_dtypes(
+    {"2.12.0 and below": ("bfloat16", "float64", "float16")}, backend_version
+)
```
Comment (ZiadAmerr) [inline ivy/functional/backends/tensorflow/experimental/layers.py:449]: I think you have removed the `with_unsupported_dtypes` from the following function, `dct` and placed it for `pool`. Can you please fix that?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py:1]: Can we refactor `_pool_args(...)` for better readability and reusability? Thanks! :)

**Review by ZiadAmerr — CHANGES_REQUESTED** [review]
Just these final comments and we are ready to merge :)

**Comment by mahmodKhaled** [comment]
Hello @ZiadAmerr, I have made some refactoring changes to ```_pool_args``` and returned again deleted ```with_unsupported_dtypes``` to ```dct``` function, please can you review the changes. if you have any comments, I'm more than welcome to apply.

**Comment by mahmodKhaled** [comment]
Hello @ZiadAmerr, I want to tell you some notes about my refactoring process to be aware of:
1. for the following three lines regarding data_formats
```
dims = draw(st.integers(min_value=1, max_value=3))
data_formats = ["NWC", "NHWC", "NDHWC"]
data_format = data_formats[dims - 1]
```
the dims variable here represents the number of dimensions in the input tensor, it starts from 1 to 3, so for example if we have an input tensor of 3 dimensions so this has to take ```NWC``` data format, so for that reason I'm subtracting 1 from dims.

Now I have refactored these three lines to be like this
```
dims = draw(st.integers(min_value=3, max_value=5))
data_formats = {3: "NWC", 4: "NHWC", 5: "NDHWC"}
data_format = data_formats[dims]
```
I make dims value start from 3 instead of 1, and max of 5 instead of 3. These new values represent the actual Input tensor shapes and make a dictionary for data_formats, that just matches every data format with its corresponding input tensor shape

2. As a result of the previous step I have replaced every line that contains ``` dims + 2``` and make it just ```dims```
3. Final concern is for ```helpers.arrays_for_pooling``` parameters which are ```min_dims``` and ```max_dims```, these parameters I have passed to them the same ```dims``` value above, for example, if input tensor shape is 3, so these parameters will have the same number 3, you may be wondering why I'm assigning the same value of ```dims``` to ```min_dims``` and ```max_dims``` and why not different values ?
So, the reason is that this line ```draw(st.integers(min_value=3, max_value=5))``` may choose value 3 to be for ```dims```, so this will take certain data format which is ```NWC```, so while generating arrays from the helper function they have to be of 3 dimensions only to be matched with the data format, if we for example, assigned ```min_dims=3``` and make ```max_dims=5``` this will generate many arrays with different dimensions ranging from 3 to 4 to 5, so all these arrays will be assigned the only data format that is ```NWC``` which will be relevant to an array of dimensions 3 and irrelevant for the others, which may make some errors. So, I decided to make these two parameters the same value :). Thank you for your efforts throughout the PR.

**Review by ZiadAmerr — APPROVED** [review]
Amazing work, @mahmodKhaled!

**Pull request merged**